### PR TITLE
Implement SMB Protocol

### PR DESCRIPTION
**Pull request by ajholland**
This implements the ability for Zgrab to scan the SMB protocol and determine if SMBv1 is supported.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -260,6 +264,16 @@ func init() {
 		zlog.Fatal("Must specify one of --tls or --starttls for --heartbleed")
 	}
 
+	// Validate SMB
+	if config.SMB.SMB {
+		if config.SMB.Protocol == 0 {
+			zlog.Fatal("Must specify protocol version for an SMB scan")
+		}
+		if config.SMB.Protocol < 0 || config.SMB.Protocol > 3 {
```
Comment (dadrian) [inline main.go]: Don't we actually only support V1?

File: ztools/smb/encoder/unicode.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	err := binary.Read(bytes.NewReader(d), binary.LittleEndian, &s)
+	if err != nil {
+		return "", err
+	}
+	return string(utf16.Decode(s)), nil
+}
+
+func ToUnicode(s string) []byte {
+	// Credit to https://github.com/Azure/go-ntlmssp/blob/master/unicode.go for logic
+	uints := utf16.Encode([]rune(s))
```
Comment (dadrian) [inline ztools/smb/encoder/unicode.go]: Out of curiosity, what does `ut16.Encode()` actually return?

File: ztools/smb/session.go [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+			return err
+		}
+		if authResp.Status != StatusOk {
+			status, _ := StatusMap[authResp.Status]
+			return errors.New(fmt.Sprintf("NT Status Error: %s\n", status))
+		}
+		s.IsAuthenticated = true
+
+		s.Debug("Completed NegotiateProtocol and SessionSetup", nil)
+	*/
```
Comment (dadrian) [inline ztools/smb/session.go]: Probably want to get rid of this commented out section

**Comment by ajholland** [comment]
@dadrian This is ready for another pass.

**Pull request merged**